### PART ONE: RECOMMENDED TO RUN ON THE CLUSTER ###
########################################################################################################
### PACKAGE INSTALLATION ###
If you are working on Adey lab servers, consider adding a pre-assembled library with dependencies to your path:

```{r}
.libPaths('/home/groups/ravnica/env/R_libs/R-4.2.2_amethyst')
```

First, load the package using devtools. 

```{r}
devtools::install_github("lrylaarsdam/amethyst")
library(amethyst)
```

########################################################################################################
### LOADING PRACTICE DATA ###

This vignette comes with a toy dataset of 50 cells derived from human brain cortex. Specify the desired path to store the file. It is 12 GB, since it is high coverage and includes both CG and CH methylation data.

```{r}
# This is the path to the vignette h5 file if using Adey Lab servers
path <- "/home/groups/ravnica/projects/amethyst/sciMETv2_braindata/brain_vignette.h5" 

# To download from AWS. Modify the desired path in the second command if you don't want it in the working directory.
download.file("https://adeylabopen.s3.us-west-2.amazonaws.com/amethyst/brain_vignette.h5", "brain_vignette.h5", method = "curl") 
path <- "./brain_vignette.h5"
```

Alternatively, many of the commands can be run using just the pre-assembled vignette workspace without the large h5 file. If you download the workspace, all the downstream commands have been run, so you can just explore what is helpful. This is recommended if you don't have cluster access.

```{r}
# This is the path to the vignette workspace if using Adey Lab servers
load("/home/groups/ravnica/projects/amethyst/sciMETv2_braindata/brain_vignette_workspace.RData") 

# Or you can download from AWS. Modify the desired path in the second command if you don't want it in the working directory.
download.file("https://adeylabopen.s3.us-west-2.amazonaws.com/amethyst/brain_vignette_workspace.RData", "brain_vignette_workspace.RData", method = "curl") 
load("./brain_vignette_workspace.RData")
```

########################################################################################################
### ASSEMBLING THE AMETHYST OBJECT ###

Now construct an amethyst object, which stores the path to the h5 file as well as a lot of other information that will be calculated downstream.

```{r}
obj <- createObject()
```

Next, we need to add metadata about each cell. Useful metadata includes quality control metrics contained in the .cellInfo.txt intermediate output or .annot files, if using the Adey lab workflow.

```{r}
obj <- addCellInfo(obj, file = "~/Downloads/brain_vignette_cellInfo.txt")
obj <- addAnnot(obj, file = "~/Downloads/brain_vignette.annot", name = "method") 
```

You can now view the information we added to the metadata like this:

```{r}
View(obj@metadata)
```

While not essential, it can be helpful to filter cells right away so downstream functions don't perform calculations
For cells that will not be used. This can easily be done with dplyr logic.

```{r}
obj@metadata <- obj@metadata |> dplyr::filter(cov > 1000000 & mch_pct < 12)
```

Next, we need to specify the location of the h5 file containing site-level methylation data for each barcode. In this case, every barcode belongs to the same h5 file, but an unlimited number of h5 files can be used in the same object. Row names are barcodes and the "paths" column specifies each path.

```{r}
obj@h5paths <- data.frame(row.names = rownames(obj@metadata), paths = rep(path, length(rownames(obj@metadata))))
```

The next step is to cluster cells, which we typically do based on methylation values over fixed genomic windows. 
An initial indexing helps reduce the computational load by determining the locations corresponding to each chromosome in every h5 file and only calculating across one at a time.
*Note: This step is recommended to run on a cluster.

```{r}
obj@index[["chr_cg"]] <- indexChr(obj, type = "CG", threads = 20)
obj@genomeMatrices[["cg_100k_score"]] <- makeWindows(obj, stepsize = 100000, type = "CG", metric = "score", threads = 20, index = "chr_cg", nmin = 5, species = "hg38")
obj@index[["chr_ch"]] <- indexChr(obj, type = "CH", threads = 20)
obj@genomeMatrices[["ch_100k_pct"]] <- makeWindows(obj, stepsize = 100000, type = "CH", metric = "percent", threads = 20, species = "hg38", index = "chr_ch", nmin = 10)
```

Next, perform dimensionality reduction with irlba. If you are unsure how many dimensions to use, the dimEstimate function can estimate the number needed to explain the desired variance threshold.
*Note: In this example, the number of requested output dimensions is low because brain_vignette.h5 has 50 cells. 

```{r}
dimEstimate(obj, genomeMatrices = c("ch_100k_pct", "cg_100k_score"), dims = c(10, 10), threshold = 0.98)
obj <- runIrlba(obj, genomeMatrices = c("ch_100k_pct", "cg_100k_score"), dims = c(5, 5))
```

Now determine cluster membership using the Rphenograph package.
*Note: In this example, k_phenograph and neighbors parameters are low because brain_vignette.h5 has 50 cells.

```{r}
obj <- regressCovBias(obj)
obj <- runCluster(obj, k_phenograph = 5) # consider increasing k_phenograph to 50 for larger datasets
obj <- runUmap(obj, neighbors = 5, dist = 0.01, method = "euclidean") 
```

Optional: one useful approach for cell type ID is analyzing CG hypomethylation over promoter regions of key marker genes. While still on the cluster, it can be helpful to calculate these cluster tracks.
*Note: This step requires a bed file of the genome divided into 500 bp windows. If using Adey lab servers, the path to the file is hardcoded. If not, the bed parameter should be the path to this file.

```{r}
obj@genomeMatrices[["cluster_cg_500_slidingwindows"]] <- makeClusterTracks(obj, type = "CG", threads = 20, bed = "hg38", index = "chr_cg") 
```

########################################################################################################
### OPTIONAL - GENE INDEXING ### 
This is only necessary if you want to look at gene-specific methylation information.
First, fetch the reference annotation file. This contains the start/end position of each gene in the genome. We need the annotation file because we want to index which components of the base .h5 file correspond to which gene. That way it can be read from a lot faster.
 
```{r}
obj@ref <- makeRef(ref = "hg38") # mm10 is also an option
```

We want hg38 in this case because our example data is from human brain. 
Indexing all the genes can take a long time, so a pre-determined list of marker genes can be loaded with the fetchMarkers command:

```{r}
markerGenes <- fetchMarkers(ref = "hg38", type = "brain") # pbmc will be added eventually
markerGenes <- c("SATB2", "GAD1") # just do a small subset for this tutorial
```

Now get the positions in the h5 file that correspond to these genes using the index function.

```{r}
index <- list() # make an empty list
index[["CH"]] <- indexGenes(hdf5 = path, gtf = obj@ref, type = "CH", subset = markerGenes, threads = 10)
index[["promoters"]] <- indexGenes(hdf5 = path, gtf = obj@ref, type = "promoters", subset = markerGenes, threads = 10) # promoter indexing finds CG base level information +/- 1500 bp around the first position of the gene (strand-aware)
```

Any type of methylation context can be indexed, as long as it is a group in the h5 file. (Example: mCAC)

Before moving to the local Rstudio, it is helpful to find cluster markers and make a matrix of % mCH over indexed genes while you can still multi-thread. 

```{r}
obj@genomeMatrices[["gene_ch"]] <- makeWindows(obj, genes = names(obj@index[["CH"]]), threads = 20, type = "CH", nmin = 2)
obj <- impute(obj, matrix = "gene_ch") # Imputation with Rmagic 
clusterMarkersCH <- findClusterMarkers(obj, matrix = "gene_ch", genes = rownames(obj@genomeMatrices[["gene_ch"]]), threads = 10) # Matrix does not have to be pre-computed, but it speeds things up downstream 
topmarkers <- clusterMarkersCH %>% dplyr::filter(p.adj < 0.05) %>% dplyr::group_by(cluster_id, direction) %>% top_n(5, wt = p.adj)
```

Now save the environment and import to RStudio, if you aren't already using it.

```{r}
save.image("brain_vignette.RData")
```

########################################################################################################
### PART TWO: LOCAL RSTUDIO ####

Load your environment saved from the server. Re-define the hdf5 path and load in amethyst dependencies. This assumes everything is in downloads; change paths if necessary.

```{r}
load("~/Downloads/brain_vignette.RData")
```

### Visualizing the results ###
First, plot the UMAP coordinates of the cells with the color corresponding to cluster membership.

```{r}
umapFeature(obj, colorBy = cluster_id)
```

umapFeature uses ggplot logic, so you can adjust as needed. For example:

```{r}
umapFeature(obj, colorBy = cluster_id, colors = sample(pal)) + facet_wrap(vars(batch)) # Batch is made up to illustrate function utility
```

umapFeature is useful for looking at how the different parameters in the cellInfo file are distributed throughout the UMAP:

```{r}
umapFeature(obj, colorBy = log(cov)) + scale_color_gradientn(colors = c("black", "turquoise", "gold", "red")) # no coverage bias!
umapFeature(obj, colorBy = mcg_pct) + scale_color_gradientn(colors = c("black", "turquoise", "gold", "red")) 
umapFeature(obj, colorBy = log(mch_pct)) + scale_color_gradientn(colors = c("black", "turquoise", "gold", "red")) # helps distinguish neuron vs glia populations
```

Show the distribution of cluster membership between samples with sampleComp. Make sure the parameters of interest are categorical variables in the metadata. Plots can be easily modified with ggplot command logic.

```{r}
sampleComp(obj, groupBy = "batch", colorBy = "cluster_id", colors = pal) 
```

########################################################################################################
### INVESTIGATING METHYLATION OVER SPECIFIC GENES ### 

Now let's look at methylation levels over key marker genes in relation to the UMAP.
Look at sigmarkers results to explore other genes.

```{r}
umapGeneM(obj, genes = c("GAD1", "SATB2", "LINGO1", "SLC17A8"), matrix = "gene_ch") 
umapGeneM(obj, genes = c("GAD1", "SATB2"), matrix = "gene_ch", blend = T) # these two are pretty mutually exclusive

```

If matrices haven't been calculated, this function can also calculate from the h5 file, but it may take longer.

```{r}
umapGeneM(obj, genes = c("GAD1", "SATB2", "LINGO1", "SLC17A8"), type = "CH") # You need "~/Downloads/brain_vignette.h5" to exist for this, or modify obj@h5path 
```

Visualization can also be easily done with violin plots or dot plots

```{r}
vlnGeneM(obj, genes = c("SATB2", "GAD1"), groupBy = cluster_id, nrow = 2, matrix = "gene_ch")
dotGeneM(obj, genes = c("SATB2", "GAD1"), type = "CH", groupBy = cluster_id) # You need "~/Downloads/brain_vignette.h5" to exist for this, or modify obj@h5path 
```

Since single-cell methylation data can be low coverage, it may be helpful to impute values.
If having trouble downloading Rmagic, try devtools::install_github("KrishnaswamyLab/MAGIC/Rmagic")

```{r}
obj <- impute(obj, matrix = "gene_ch") # Already added if using /home/groups/ravnica/projects/amethyst/sciMETv2_braindata/brain_vignette_workspace.RData
```

Then use the resulting matrix to visualize from imputed values

```{r}
p1 <- umapGeneM(obj, genes = c("SLC17A8"), matrix = "gene_ch")
p2 <- umapGeneM(obj, genes = c("SLC17A8"), matrix = "gene_ch_imputed") 
plot_grid(p1, p2)
```

Visualization of methylation over the gene body can be done with histGeneM or tileGeneM

```{r}
histGeneM(obj, type = "CH", groupBy = cluster_id, genes = c("GAD1", "LINGO1"), bins = 40) # You need "~/Downloads/brain_vignette.h5" to exist for this, or modify obj@h5path 
tileGeneM(obj, genes = c("MOG", "C1QA", "GAD1", "SLC17A7", "SST", "AQP4"), matrix = "cluster_cg_500_slidingwindows", nrow = 3) 
```

Variation in methylation levels of key marker genes can help aid in cell type ID, but it may be more effective to compare to a reference. If using brain data you can use clusterCompare to correlate % methylation over 100kb genomic windows to a reference generated from Liu 2021 (Ecker lab). Levels to compare with are:
"region_name"  "major_region" "sub_region"   "cell_class"   "major_type"   "sub_type" 

First, generate an aggregated matrix of values over some metadata parameter.

```{r}
obj <- aggregateMatrix(obj, matrix = "ch_100k_pct", groupBy = cluster_id, name = "cluster_ch_100k_pct") # Already added if using /home/groups/ravnica/projects/amethyst/sciMETv2_braindata/brain_vignette_workspace.RData
clusterCompare("cluster_ch_100k_pct", level = "major_type", type = "ch", n = 5) # Correlation very low because it is human vs. mouse. Still working on this
```

Coming up:
- Trajectory analysis
- DMR identification
- Open to suggestions



testPalette("umap", obj = obj)



